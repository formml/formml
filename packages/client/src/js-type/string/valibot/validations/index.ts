import bool from './bool.js'
import datetime from './datetime.js'
import decimal from './decimal.js'
import num from './num.js'

export { bool, datetime, decimal, num }
