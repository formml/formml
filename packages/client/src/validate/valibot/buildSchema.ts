import { Field } from '@formml/dsl'

export default function buildSchema(formmlSchema: Field) {
  return formmlSchema
}
