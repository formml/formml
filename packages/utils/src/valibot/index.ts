export * from './schemas/required.js'
export * from './validations/isoDateTime.js'
export * from './validations/numerical.js'
export * from './validations/allowBlank.js'
export * from './validations/numeric-validations.js'
