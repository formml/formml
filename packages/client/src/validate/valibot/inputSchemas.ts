import * as v from 'valibot'

export const number = () => v.special<string>((input) => !isNaN(Number(input)))
