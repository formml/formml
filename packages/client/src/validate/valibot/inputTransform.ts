import { BigNumber } from 'bignumber.js'
import dayjs from 'dayjs'
import * as v from 'valibot'

export const toNum = (schema?: v.BaseSchema) =>
  v.transform<v.StringSchema, number>(
    v.string([v.custom((i) => !isNaN(Number(i)))]),
    Number,
    schema,
  )
export const toDatetime = (schema?: v.BaseSchema) =>
  v.transform<v.StringSchema, dayjs.Dayjs>(
    v.string([v.isoTimestamp()]),
    dayjs,
    schema,
  )
export const toBool = (schema?: v.BaseSchema) =>
  v.transform<v.StringSchema, boolean>(v.string(), Boolean, schema)
export const toDecimal = (schema?: v.BaseSchema) =>
  v.transform<v.StringSchema, BigNumber>(
    v.string([v.custom((i) => !new BigNumber(i).isNaN())]),
    (i) => new BigNumber(i),
    schema,
  )
