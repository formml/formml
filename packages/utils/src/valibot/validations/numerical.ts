import type { BaseIssue, BaseValidation, Dataset, ErrorMessage } from 'valibot'

import { _addIssue } from 'valibot'

export interface NumericalIssue<TInput extends string>
  extends BaseIssue<TInput> {
  /**
   * The expected property.
   */
  readonly expected: 'numerical'
  /**
   * The issue kind.
   */
  readonly kind: 'validation'
  /**
   * The received property.
   */
  readonly received: `"${string}"`
  /**
   * The validation function.
   */
  readonly requirement: (value: string) => boolean
  /**
   * The issue type.
   */
  readonly type: 'numerical'
}

export interface NumericalAction<
  TInput extends string,
  TMessage extends ErrorMessage<NumericalIssue<TInput>> | undefined,
> extends BaseValidation<TInput, TInput, NumericalIssue<TInput>> {
  /**
   * The expected property.
   */
  readonly expects: 'numerical'
  /**
   * The error message.
   */
  readonly message: TMessage
  /**
   * The action reference.
   */
  readonly reference: typeof numerical
  /**
   * The validation function.
   */
  readonly requirement: (value: string) => boolean
  /**
   * The action type.
   */
  readonly type: 'numerical'
}

/**
 * Creates a numerical validation action.
 *
 * @returns A numerical action.
 */
export function numerical<TInput extends string>(): NumericalAction<
  TInput,
  undefined
>

/**
 * Creates a numerical validation action.
 *
 * @param message The error message.
 *
 * @returns A numerical action.
 */
export function numerical<
  TInput extends string,
  const TMessage extends ErrorMessage<NumericalIssue<TInput>> | undefined,
>(message: TMessage): NumericalAction<TInput, TMessage>

export function numerical(
  message?: ErrorMessage<NumericalIssue<string>>,
): NumericalAction<string, ErrorMessage<NumericalIssue<string>> | undefined> {
  return {
    _run(dataset, config) {
      if (dataset.typed && !this.requirement(dataset.value)) {
        _addIssue(this, 'input', dataset, config)
      }
      return dataset as Dataset<string, NumericalIssue<string>>
    },
    async: false,
    expects: 'numerical',
    kind: 'validation',
    message,
    reference: numerical,
    requirement: (value) => value.trim() !== '' && !isNaN(Number(value)),
    type: 'numerical',
  }
}
