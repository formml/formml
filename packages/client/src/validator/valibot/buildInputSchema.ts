import { Field } from '@formml/dsl'

import { assertNever } from '../../utils/assertNever.js'
import buildSchema from './buildSchema.js'
import * as t from './transform/index.js'

export default function buildInputSchema(formmlSchema: Field) {
  if (formmlSchema.type === 'text') {
    const typedSchema = buildSchema(formmlSchema)
    return typedSchema
  }

  if (formmlSchema.type === 'num') {
    const typedSchema = buildSchema(formmlSchema)
    return t.toNum(typedSchema)
  }

  if (formmlSchema.type === 'bool') {
    const typedSchema = buildSchema(formmlSchema)
    return t.toBool(typedSchema)
  }

  if (formmlSchema.type === 'datetime') {
    const typedSchema = buildSchema(formmlSchema)
    return t.toDatetime(typedSchema)
  }

  if (formmlSchema.type === 'decimal') {
    const typedSchema = buildSchema(formmlSchema)
    return t.toDecimal(typedSchema)
  }

  return assertNever`Unsupported field schema: ${formmlSchema}`
}
