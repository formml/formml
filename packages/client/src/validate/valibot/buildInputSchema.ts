import { Field } from '@formml/dsl'

import { assertNever } from '../../utils/assertNever.js'
import buildSchema from './buildSchema.js'
import * as i from './inputTransform.js'

export default function buildInputSchema(formmlSchema: Field) {
  if (formmlSchema.type === 'text') {
    return buildSchema(formmlSchema)
  }

  if (formmlSchema.type === 'num') {
    return i.toNum()
  }

  if (formmlSchema.type === 'bool') {
    return i.toBool()
  }

  if (formmlSchema.type === 'datetime') {
    return i.toDatetime()
  }

  if (formmlSchema.type === 'decimal') {
    return i.toDecimal()
  }

  return assertNever`Unsupported field type: ${formmlSchema.type}`
}
