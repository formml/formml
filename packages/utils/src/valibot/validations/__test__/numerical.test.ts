import * as v from 'valibot'

import { numerical } from '../numerical.js'

describe('numerical', () => {
  test('should validate a numerical string', () => {
    // Arrange
    const schema = v.pipe(v.string(), numerical())

    // Act
    const result = v.safeParse(schema, '123')

    // Assert
    expect(result.success).toBe(true)
  })

  test.each(['Infinity', '-Infinity'])(
    'should validate infinities - "%s"',
    (input) => {
      // Arrange
      const schema = v.pipe(v.string(), numerical())

      // Act
      const result = v.safeParse(schema, input)

      // Assert
      expect(result.success).toBe(true)
    },
  )

  test.each(['', '  ', ' \n\t'])(
    'should invalidate a blank string',
    (input) => {
      // Arrange
      const schema = v.pipe(v.string(), numerical())

      // Act
      const result = v.safeParse(schema, input)

      // Assert
      expect(result.success).toBe(false)
      expect(result.issues).toMatchSnapshot()
    },
  )

  test('should invalidate a non-numerical string', () => {
    // Arrange
    const schema = v.pipe(v.string(), numerical())

    // Act
    const result = v.safeParse(schema, 'abc')

    // Assert
    expect(result.success).toBe(false)
    expect(result.issues).toMatchInlineSnapshot(`
      [
        {
          "abortEarly": undefined,
          "abortPipeEarly": undefined,
          "expected": "numerical",
          "input": "abc",
          "issues": undefined,
          "kind": "validation",
          "lang": undefined,
          "message": "Invalid input: Expected numerical but received "abc"",
          "path": undefined,
          "received": ""abc"",
          "requirement": [Function],
          "type": "numerical",
        },
      ]
    `)
  })

  test('should invalidate NaN', () => {
    // Arrange
    const schema = v.pipe(v.string(), numerical())

    // Act
    const result = v.safeParse(schema, 'NaN')

    // Assert
    expect(result.success).toBe(false)
    expect(result.issues).toMatchInlineSnapshot(`
      [
        {
          "abortEarly": undefined,
          "abortPipeEarly": undefined,
          "expected": "numerical",
          "input": "NaN",
          "issues": undefined,
          "kind": "validation",
          "lang": undefined,
          "message": "Invalid input: Expected numerical but received "NaN"",
          "path": undefined,
          "received": ""NaN"",
          "requirement": [Function],
          "type": "numerical",
        },
      ]
    `)
  })

  test('should accept custom message', () => {
    // Arrange
    const schema = v.pipe(v.string(), numerical('Custom message'))

    // Act
    const result = v.safeParse(schema, 'abc')

    // Assert
    expect(result.success).toBe(false)
    expect(result.issues).toMatchInlineSnapshot(`
      [
        {
          "abortEarly": undefined,
          "abortPipeEarly": undefined,
          "expected": "numerical",
          "input": "abc",
          "issues": undefined,
          "kind": "validation",
          "lang": undefined,
          "message": "Custom message",
          "path": undefined,
          "received": ""abc"",
          "requirement": [Function],
          "type": "numerical",
        },
      ]
    `)
  })
})
