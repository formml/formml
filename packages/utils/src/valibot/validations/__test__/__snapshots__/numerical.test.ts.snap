// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`numerical > should invalidate a blank string 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "numerical",
    "input": "",
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid input: Expected numerical but received """,
    "path": undefined,
    "received": """",
    "requirement": [Function],
    "type": "numerical",
  },
]
`;

exports[`numerical > should invalidate a blank string 2`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "numerical",
    "input": "  ",
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid input: Expected numerical but received "  "",
    "path": undefined,
    "received": ""  "",
    "requirement": [Function],
    "type": "numerical",
  },
]
`;

exports[`numerical > should invalidate a blank string 3`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "expected": "numerical",
    "input": " 
	",
    "issues": undefined,
    "kind": "validation",
    "lang": undefined,
    "message": "Invalid input: Expected numerical but received " 
	"",
    "path": undefined,
    "received": "" 
	"",
    "requirement": [Function],
    "type": "numerical",
  },
]
`;
